### PR TITLE
net/tcp/udp: allocate tcp/udp connect dynamically 

### PR DESCRIPTION
**Pull request by anchao**
## Summary

net/udp: rename NET_UDP_CONNS to NET_UDP_CONNS_PER_ALLOC
net/tcp: rename NET_TCP_CONNS to NET_TCP_CONNS_PER_ALLOC
net/udp: allocate udp connect dynamically
net/tcp: allocate tcp connect dynamically

## Impact

tcp/udp connect instance allocate

## Testing

tcp/udp client/server test

### CODE REVIEW

File: net/tcp/tcp_conn.c [hunk 1]
```diff
@@ -61,6 +61,8 @@
 #include <nuttx/net/ip.h>
 #include <nuttx/net/tcp.h>
 
+#include <nuttx/kmalloc.h>
```
Comment (xiaoxiang781216) [inline net/tcp/tcp_conn.c:64]: move before line 57

File: net/tcp/tcp_conn.c [hunk 2]
```diff
@@ -656,8 +677,6 @@ FAR struct tcp_conn_s *tcp_alloc(uint8_t domain)
     }
 #endif
 
-  net_unlock();
```
Comment (xiaoxiang781216) [inline net/tcp/tcp_conn.c:659]: let's move tcp_alloc_conn after net_unlock?

File: include/nuttx/net/netconfig.h [hunk 3]
```diff
@@ -456,11 +456,11 @@
  * connection requires approximately 30 bytes of memory.
  */
 
-#ifndef CONFIG_NET_TCP_CONNS
+#ifndef CONFIG_NET_TCP_CONNS_PER_BLOCK
 #  ifdef CONFIG_NET_TCP
-#   define CONFIG_NET_TCP_CONNS 10
+#   define CONFIG_NET_TCP_CONNS_PER_BLOCK 10
```
Comment (davids5) [inline include/nuttx/net/netconfig.h:461]: The new default in Kconfig is `default 4`  is it better these this match?

File: net/tcp/tcp_conn.c [hunk 4]
```diff
@@ -110,15 +108,12 @@ static FAR struct tcp_conn_s *
   tcp_listener(uint8_t domain, FAR const union ip_addr_u *ipaddr,
                uint16_t portno)
 {
-  FAR struct tcp_conn_s *conn;
-  int i;
+  FAR struct tcp_conn_s *conn = NULL;
```
[the remark this answers is not in the document]
Comment (anchao) [reply net/tcp/tcp_conn.c:111]: NULL indicates to get the active list from tcp_nextconn()

File: net/tcp/tcp_conn.c (hunk 1 above)
Comment (anchao) [reply net/tcp/tcp_conn.c:64]: Done

File: net/tcp/tcp_conn.c (hunk 2 above)
Comment (anchao) [reply net/tcp/tcp_conn.c:659]: Done

File: include/nuttx/net/netconfig.h (hunk 3 above)
Comment (anchao) [reply include/nuttx/net/netconfig.h:461]: Done

File: net/tcp/tcp_conn.c [hunk 5]
```diff
@@ -538,22 +570,10 @@ static inline int tcp_ipv6_bind(FAR struct tcp_conn_s *conn,
...
   dq_init(&g_free_tcp_connections);
   dq_init(&g_active_tcp_connections);
-
-  /* Now initialize each connection structure */
-
-  for (i = 0; i < CONFIG_NET_TCP_CONNS; i++)
-    {
-      /* Mark the connection closed and move it to the free list */
-
-      g_tcp_connections[i].tcpstateflags = TCP_CLOSED;
```
Comment (davids5) [inline net/tcp/tcp_conn.c:554]: Where is this functionality now?

Comment (anchao) [reply net/tcp/tcp_conn.c:554]: line 548 - 550
https://github.com/apache/incubator-nuttx/pull/5104/commits/6038a35bc5c57d17c6ab277906d5c3e696a43c27#diff-fd66b1e5b4be7133f600f2a9089d2415165ef1fd8ff957e8aef261b119f7394eR548-R550

**Review by davids5 — APPROVED** [review]
LGTM

File: net/udp/udp_conn.c [hunk 6]
```diff
@@ -62,6 +62,8 @@
 #include <nuttx/net/ip.h>
 #include <nuttx/net/udp.h>
 
+#include <nuttx/kmalloc.h>
```
Comment (xiaoxiang781216) [inline net/udp/udp_conn.c:65]: move before nuttx/clock.h too

File: net/tcp/Kconfig [hunk 7]
```diff
@@ -50,11 +50,11 @@ config NET_TCPURGDATA
 		compiled in. Urgent data (out-of-band data) is a rarely used TCP feature
 		that is very seldom would be required.
 
-config NET_TCP_CONNS
-	int "Number of TCP/IP connections"
-	default 8
+config NET_TCP_CONNS_PER_BLOCK
+	int "Number of TCP/IP connections per block(one for each open)"
+	default 4
```
Comment (xiaoxiang781216) [inline net/tcp/Kconfig:55]: let's default to 1 when DEFAULT_SMALL is true

File: net/udp/Kconfig [hunk 8]
```diff
@@ -50,11 +50,11 @@ config NET_UDP_CHECKSUMS
 		Enable/disable UDP checksum support.  UDP checksum support is
 		REQUIRED for IPv6.
 
-config NET_UDP_CONNS
-	int "Number of UDP sockets"
-	default 8
+config NET_UDP_CONNS_PER_BLOCK
+	int "Number of UDP sockets per block(one for each open)"
+	default 4
```
Comment (xiaoxiang781216) [inline net/udp/Kconfig:55]: default to 1 in the small config

File: net/udp/Kconfig [hunk 9]
```diff
@@ -50,11 +50,11 @@ config NET_UDP_CHECKSUMS
...
 	---help---
-		The maximum amount of open concurrent UDP sockets
+		The maximum amount of open concurrent UDP sockets per block
```
Comment (xiaoxiang781216) [inline net/udp/Kconfig:57]: The amount of open concurrent UDP sockets per allocation

File: net/tcp/Kconfig [hunk 10]
```diff
@@ -50,11 +50,11 @@ config NET_TCPURGDATA
...
 	---help---
-		Maximum number of TCP/IP connections (all tasks)
+		Maximum number of TCP/IP connections per block(all tasks)
```
Comment (xiaoxiang781216) [inline net/tcp/Kconfig:57]: The  number of TCP/IP connections per allocation(all tasks)

Comment (davids5) [reply net/tcp/Kconfig:57]: What aboutn NET_{UDP|TCP}_CONNS_PER_BLOCK -> NET_{UDP|TCP}_CONNS_PER_ALLOC  so it is consistent with this behavior that is not the same as the FD change that is blocks?

File: net/udp/Kconfig (hunk 9 above)
Comment (anchao) [reply net/udp/Kconfig]: Done

File: net/udp/Kconfig (hunk 8 above)
Comment (anchao) [reply net/udp/Kconfig:56]: Done

File: net/tcp/Kconfig (hunk 10 above)
Comment (anchao) [reply net/tcp/Kconfig]: Done

File: net/tcp/Kconfig (hunk 7 above)
Comment (anchao) [reply net/tcp/Kconfig:56]: Done

File: net/udp/udp_conn.c (hunk 6 above)
Comment (anchao) [reply net/udp/udp_conn.c]: Done

File: net/udp/Kconfig [hunk 11]
```diff
@@ -50,11 +50,12 @@ config NET_UDP_CHECKSUMS
...
-config NET_UDP_CONNS
-	int "Number of UDP sockets"
-	default 8
+config NET_UDP_CONNS_PER_ALLOC
+	int "Number of UDP sockets per block(one for each open)"
+	default 1 if DEFAULT_SMALL
+	default 4
 	---help---
-		The maximum amount of open concurrent UDP sockets
+		The amount of open concurrent UDP sockets per allocation
```
Comment (davids5) [inline net/udp/Kconfig:58]: The number of UDP sockets allocated per allocation that occurs when none are available.

File: net/udp/Kconfig [hunk 12]
```diff
@@ -50,11 +50,12 @@ config NET_UDP_CHECKSUMS
 		Enable/disable UDP checksum support.  UDP checksum support is
 		REQUIRED for IPv6.
 
-config NET_UDP_CONNS
-	int "Number of UDP sockets"
-	default 8
+config NET_UDP_CONNS_PER_ALLOC
+	int "Number of UDP sockets per block(one for each open)"
```
Comment (davids5) [inline net/udp/Kconfig:54]: Number of UDP sockets per allocation.

**Review by davids5 — CHANGES_REQUESTED** [review]
Just some word wordsmithing needed.

File: net/udp/Kconfig (hunk 11 above)
Comment (anchao) [reply net/udp/Kconfig]: Done

File: net/udp/Kconfig (hunk 12 above)
Comment (anchao) [reply net/udp/Kconfig]: Done

**Review by davids5 — COMMENTED** [review]
@anchao Please grep the diff for `block` and change it to allocation everywhere.

**Review by davids5 — APPROVED** [review]
@anchao Thank you!

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by gustavonihei** [comment]
What's the motivation for allocating the resources dynamically?
The decision for this type of change should consider an analysis of the tradeoffs.

File: net/tcp/tcp_conn.c [hunk 13]
```diff
@@ -110,15 +107,12 @@ static FAR struct tcp_conn_s *
...
                uint16_t portno)
 {
-  FAR struct tcp_conn_s *conn;
-  int i;
+  FAR struct tcp_conn_s *conn = NULL;
 
   /* Check if this port number is in use by any active UIP TCP connection */
 
-  for (i = 0; i < CONFIG_NET_TCP_CONNS; i++)
+  while ((conn = tcp_nextconn(conn)) != NULL)
```
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply net/tcp/tcp_conn.c:114]: Should we protect DoS from the server level?

[the remark this answers is not in the document]
Comment (acassis) [reply net/tcp/tcp_conn.c:114]: I think so! Also imagine a board connected to a Wi-Fi router, anyone in the same network could explore it.

**Comment by acassis** [comment]
I think dynamic allocation in the kernel makes NuttX more unpredictable for realtime applications. Maybe it could be dependent of some option like CONFIG_ALLOW_NON_RT_BEHAVIOR. What do you think?

**Comment by xiaoxiang781216** [comment]
> What's the motivation for allocating the resources dynamically? The decision for this type of change should consider an analysis of the tradeoffs.

To support the complex system better, we have saw in a production system:

1. Application or server may create more socket in some special case
2. New version may create more socket and then old config stop working
3. TIME_WAIT state in TCP make the situation even worse:
    http://www.serverframework.com/asynchronousevents/2011/01/time-wait-and-its-design-implications-for-protocols-and-scalable-servers.html

Yes, it's a tradeoff, but dynamic allocation already exist inside socket API before this patch.

**Comment by xiaoxiang781216** [comment]
> I think dynamic allocation in the kernel makes NuttX more unpredictable for realtime applications.

Yes, TLSF allocator may improve the real time a little bit(http://www.gii.upv.es/tlsf/).

> Maybe it could be dependent of some option like CONFIG_ALLOW_NON_RT_BEHAVIOR. What do you think?

or we can do the initial allocation in initialization since the realtime behavior is same as before until we use up the reserved conn.

**Comment by anchao** [comment]
How about adding an dynamic enable option? If dynamic allocate is enabled on the basis of the pre-allocation pool, 
then we will allocate new connection instances from the heap.
Of course, we will also add the maximum limit config to protect the system from malicious attacks.

CONFIG_NET_DYNAMIC_CONNS
CONFIG_NET_MAX_DYNAMIC_CONNS

**Comment by anchao** [comment]
The current static configuration is no longer suitable for business situation, this is intention of pull request,
In terms of real products, NuttX is not just running in simple IOT devices,
we already have a demand for full-featured business terminal, We have online video/audio streaming services,
stream transfer services, adb network bridges, high-level application networking services, etc.
Most of the time we serve in an environment where the number of connections is unknown,
For example if multicast or mesh network device as long as there is one more hardware device connected to our service in the local area network, the number of connections will increase by one. this is common scene of smart homes.

We hope to find a balance to enhance the scalability of NuttX. In most cases, other developers will have similar situations.

This connection information is dumped from Smart Speaker, the number of connections is still growing from application developers ...

```
 [TCP]                 LADDR    LPORT             RADDR      RPORT  RCVBUFS  SNDBUFS  SNDPEND  RCVPEND   RCVOOO
 TCP   [ 0]: [       127.0.0.1: 12904] -> [       127.0.0.1: 12903]    16384    16384        0        0        0
 TCP   [ 1]: [       127.0.0.1: 12903] -> [       127.0.0.1: 12904]    16384    16384        0        0        0
 TCP   [ 2]: [         0.0.0.0: 12905] -> [       127.0.0.1: 54322]    16384    16384        0        0        0
 TCP   [ 3]: [         0.0.0.0: 12906] -> [       127.0.0.1: 54323]    16384    16384        0        0        0
 TCP   [ 4]: [       127.0.0.1: 54322] -> [       127.0.0.1: 12905]    16384    16384        0        0        0
 TCP   [ 5]: [       127.0.0.1: 54323] -> [       127.0.0.1: 12906]    16384    16384        0        0        0
 TCP   [ 6]: [         0.0.0.0: 12908] -> [ ***************:   443]    16384    16384        0        0        0
 TCP   [ 7]: [         0.0.0.0: 12924] -> [ ***************:    80]    16384    16384        0        0        0
 TCP   [ 8]: [         0.0.0.0: 13689] -> [ ***************:    80]   131070    16384        0      541        0
 TCP   [ 9]: [         0.0.0.0: 13763] -> [ ***************:   443]    16384    16384        0        0      213
 TCP   [10]: [         0.0.0.0: 14322] -> [ ***************:   645]    16384    16384        0        0        0
 TCP   [11]: [         0.0.0.0: 14323] -> [ ***************:    80]   131070    16384        0        0        0
 TCP   [12]: [         0.0.0.0: 14633] -> [ ***************:   443]    16384    16384        0        0        0
 TCP   [13]: [         0.0.0.0: 14634] -> [ ***************:   443]    16384    16384        0        0        0
 TCP   [14]: [         0.0.0.0: 14711] -> [ ***************:    80]   131070    16384        0      123        0
 -----------------------
 [UDP]                 LADDR    LPORT             RADDR      RPORT  RCVBUFS  SNDBUFS  SNDPEND  RCVPEND
 UDP   [ 0] [         0.0.0.0: 54321] -> [    36.156.49.91:   443]:    16384    16384        0        0
 UDP   [ 1] [         0.0.0.0: 21355] -> [  111.202.86.129:  6666]:    16384    16384        0        0
 UDP   [ 2] [         0.0.0.0: 10010] -> [         0.0.0.0:     0]:    16384    16384        0        0
 UDP   [ 3] [         0.0.0.0: 21356] -> [   192.168.31.93: 29443]:   262144   262144        0        0
 UDP   [ 4] [         0.0.0.0: 21357] -> [ ***************: 10010]:    16384    16384        0        0
 UDP   [ 5] [         0.0.0.0: 21340] -> [   192.168.31.96: 29443]:    16384    16384        0        0
 UDP   [ 6] [         0.0.0.0: 21341] -> [ ***************: 10010]:    16384    16384        0        0
 UDP   [ 7] [         0.0.0.0: 21345] -> [  192.168.31.133: 29443]:    16384    16384        0        0
 UDP   [ 8] [         0.0.0.0: 21365] -> [ ***************:  6432]:    16384    16384        0        0
 UDP   [ 9] [         0.0.0.0: 21377] -> [  192.168.31.147: 29443]:   262144   262144        0        0
 UDP   [10] [         0.0.0.0: 21378] -> [ ***************:  7981]:    16384    16384        0        0
```

**Comment by acassis** [comment]
> How about adding an dynamic enable option? If dynamic allocate is enabled on the basis of the pre-allocation pool, then we will allocate new connection instances from the heap. Of course, we will also add the maximum limit config to protect the system from malicious attacks.
> 
> CONFIG_NET_DYNAMIC_CONNS CONFIG_NET_DYNAMIC_CONNS_PER_ALLOC CONFIG_NET_MAX_DYNAMIC_CONNS

Yes, seams a good approach!

**Comment by anchao** [comment]
Update the request,  add new config about connec dynamic allocate:

`CONFIG_NET_ALLOC_CONNS`

if CONFIG_NET_ALLOC_CONNS is enabled, the static pre-applied connection entries will be disabled,
all connections will be dynamically allocated from heap. about the maximum number of connections,
we prepare implement the limitation from the VFS layer:

```
_POSIX_OPEN_MAX
RLIMIT_NOFILE
```

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@acassis do you think the new implementation is good to merge?

**Review by davids5 — CHANGES_REQUESTED** [review]
This is a much better approach allowing be changed.

File: net/Kconfig [hunk 14]
```diff
@@ -312,6 +312,15 @@ source "net/ipforward/Kconfig"
 
 endmenu # Internet Protocol Selection
 
+config NET_ALLOC_CONNS
+	bool "Allocate connect instance dynamically support"
```
[the remark this answers is not in the document]
Comment (anchao) [reply net/Kconfig:316]: Done

File: net/Kconfig [hunk 15]
```diff
@@ -312,6 +312,15 @@ source "net/ipforward/Kconfig"
...
+	default n
+	---help---
+		Enable the function of Allocate connect instance dynamically.
+		If the connections can not be estimated, the stack will abandon
+		static pre-allocate connection entries, all connections will be
+		dynamically allocated from heap.
+
```
[the remark this answers is not in the document]
Comment (anchao) [reply net/Kconfig:324]: Done

**Review by davids5 — APPROVED** [review]
@anchao - thank you.

**Comment by xiaoxiang781216** [comment]
@anchao please make the similar change to other implementation

**Pull request merged**